core/services/directrequestocr/drlistener.go

```diff
@@ -0,0 +1,271 @@
+package directrequestocr
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"sync"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common"
+
+	"github.com/smartcontractkit/chainlink/core/chains/evm/log"
+	"github.com/smartcontractkit/chainlink/core/gethwrappers/generated/ocr2dr_oracle"
+	"github.com/smartcontractkit/chainlink/core/logger"
+	"github.com/smartcontractkit/chainlink/core/services/job"
+	"github.com/smartcontractkit/chainlink/core/services/ocr2/plugins/directrequestocr/config"
+	"github.com/smartcontractkit/chainlink/core/services/pg"
+	"github.com/smartcontractkit/chainlink/core/services/pipeline"
+	"github.com/smartcontractkit/chainlink/core/utils"
+)
+
+var (
+	_ log.Listener   = &DRListener{}
+	_ job.ServiceCtx = &DRListener{}
+)
+
+const (
+	ParseResultTaskName string = "parse_result"
+	ParseErrorTaskName  string = "parse_error"
+)
+
+type DRListener struct {
+	oracle            ocr2dr_oracle.OCR2DROracle
+	job               job.Job
+	pipelineRunner    pipeline.Runner
+	jobORM            job.ORM
+	logBroadcaster    log.Broadcaster
+	runs              sync.Map
+	shutdownWaitGroup sync.WaitGroup
+	mbOracleEvents    *utils.Mailbox[log.Broadcast]
+	chStop            chan struct{}
+	pluginORM         ORM
+	pluginConfig      config.PluginConfig
+	logger            logger.Logger
+	utils.StartStopOnce
+}
+
+func NewDRListener(oracle ocr2dr_oracle.OCR2DROracle, jb job.Job, runner pipeline.Runner, jobORM job.ORM, pluginORM ORM, pluginConfig config.PluginConfig, logBroadcaster log.Broadcaster, lggr logger.Logger) *DRListener {
+	return &DRListener{
+		oracle:         oracle,
+		job:            jb,
+		pipelineRunner: runner,
+		jobORM:         jobORM,
+		logBroadcaster: logBroadcaster,
+		mbOracleEvents: utils.NewHighCapacityMailbox[log.Broadcast](),
+		chStop:         make(chan struct{}),
+		pluginORM:      pluginORM,
+		pluginConfig:   pluginConfig,
+		logger:         lggr,
+	}
+}
+
+// Start complies with job.Service
+func (l *DRListener) Start(context.Context) error {
+	return l.StartOnce("DirectRequestListener", func() error {
+		unsubscribeLogs := l.logBroadcaster.Register(l, log.ListenerOpts{
+			Contract: l.oracle.Address(),
+			ParseLog: l.oracle.ParseLog,
+			LogsWithTopics: map[common.Hash][][]log.Topic{
+				ocr2dr_oracle.OCR2DROracleOracleRequest{}.Topic():  {},
+				ocr2dr_oracle.OCR2DROracleOracleResponse{}.Topic(): {},
+			},
+			MinIncomingConfirmations: l.pluginConfig.MinIncomingConfirmations,
```

Comment:
would this be strict finality? if not, then you could see the same reqID twice from the lb

Reply:
Adding a check for repeated IDs.